### DISCUSSION TITLE
How to summarize the attributes from model?

### DISCUSSION THREAD

**Discussion by NaungYeHtet** [filamentphp/filament#13080] [Help]
### Package

Panel builder

### Package Version

v3

### How can we help you?

```
Tables\Columns\TextColumn::make('price')
    ->summarize([
        Tables\Columns\Summarizers\Sum::make(),
    ]),
```

```
protected function price(): Attribute
    {
        return Attribute::make(
            get: function (mixed $value, array $attributes) {
                return 100 * 344;
            },
        );
    }
```

Got sql error unknown column price

**Comment by leandrocfe** [comment]
Summarize will use the db, not attributes. If you need to calculate the price, you can use a virtual column as a table column, like this:

```php
//migration..
$table->float('price')->virtualAs('column_a * column_b');
```

**Reply by luoyou** [reply]
> Summarize will use the db, not attributes. If you need to calculate the price, you can use a virtual column as a table column, like this:
> 
> ```
> //migration..
> $table->float('price')->virtualAs('column_a * column_b');
> ```

The virtual field is difficult to meet the requirements. Is there any way to implement it through model attributes, for example, using callback injection into the model

**Reply by StarterES-sgaleano** [reply]
Haya, did you found any workaround for this? Im facing the same issue and I cant find any answer out there. 

Thanks in advance.

**Reply by luoyou** [reply]
I found a solution.
```php
use Filament\Tables\Columns\Summarizers\Summarizer;

TextColumn::make('mobile_price')
    ->formatStateUsing(fn (string $state) => $this->priceByAgent($price, $state))
    ->summarize(Summarizer::make()
        ->label('')
        ->using(fn (Table $table) => $this->priceByAgent($price, $this->mergePrice($table->getRecords()))))
```

**Reply by akki-io** [reply]
You can do

```php
use Filament\Tables\Columns\Summarizers\Summarizer;
use Filament\Tables\Table;

Tables\Columns\TextColumn::make('price')
    ->summarize(Summarizer::make()            
        ->using(function (Table $table) {
            return $table->getRecords()->sum('price');                
        })            
    ),

```

**Reply by StarterES-sgaleano** [reply]
Thanks you all, i have done that, the thing is that when you group records, the summary for the group will show the total sum instead of that of the group.
<img width="817" alt="image" src="https://github.com/user-attachments/assets/4741da58-2a33-4a2a-bfc7-59a790826b6b">

**Reply by long-blade** [reply]
yes but this i think is not working for summary of all pages! the all pages and the this page have always the same summarize!

**Comment by DL-Works** [comment]
public function getGivenFeesAmountAttribute()
    {
        return $this->payments
            ->sum('amount');
    }
    
    want to use this attribute column for summary. Has anybody found a workaround?

**Comment by RX-77IT** [comment]
I think I got it working, but **BE CAREFUL** to check with the Debugbar how many new queries you're introducing, especially when loading relationships.

```php
->summarize(
    Tables\Columns\Summarizers\Summarizer::make()
        ->label('Total label')
        ->using(function ($query) {
            $ids = $query->pluck('id'); // assuming the model is using ids
            $total = static::getModel()::whereIn('id', $ids)
                ->get()
                ->sum('attribute_name');
            return $total;
        })
)
```

Basically I'm plucking all the ids of the paginated $query (because doing a `dd($query->get())` shows that there are no attributes loaded) and then I re-query the Model to get a filtered collection using the plucked ids. 
In that collection the items now have the attribute needed and we can use sum() (from the [Collection methods](https://laravel.com/docs/12.x/collections#available-methods)) or something else.

It's a bit convoluted, but I don't know how to proceed from here and improve it.

___________


**After lunch EDIT**: I can simplify one query (the one taking all records) by checking if count($ids) is equal to one of the pagination options of the page like so.

```php
use Filament\Tables\Table;

->summarize(
    Tables\Columns\Summarizers\Summarizer::make()
        ->label('Total label')
        ->using(function ($query, Table $table) {
            $ids = $query->pluck('id'); // assuming the model is using ids
            $total = static::getModel()::query()
                ->when(
                    in_array(count($ids), $table->getPaginationPageOptions()),
                    fn ($q) => $q->whereIn('id', $ids)
                )
                ->get()
                ->sum('attribute_name');
            return $total;
        })
)
```

This way if there is a million records the whereIn() filter doesn't apply to all records.

**Reply by RX-77IT** [reply]
Another way to do it is not using the attribute, but recreate it (if possible) with $table->modifyQueryUsing() and append some aggregated data like [withSum](https://laravel.com/docs/12.x/eloquent-relationships#other-aggregate-functions) or [withCount](https://laravel.com/docs/12.x/eloquent-relationships#counting-related-models) or withAggregate().

```php
$table
  ->modifyQueryUsing(fn (\Illuminate\Database\Eloquent\Builder $query) => $query
    ->withSum(['relation as ficticious_column' => fn ($q) => $q->where('condition', true)], 'real_column_to_sum')
  )
```
```php
->summarize([
  Tables\Columns\Summarizers\Summarizer::make()
    ->label('label')
    ->using(fn ($query) => $query->sum('ficticious_column')),
])
```

Povilas did a pretty good video on this: https://www.youtube.com/watch?v=-5aTOQKkVs0
